tracing-core/src/metadata.rs

```diff
@@ -60,22 +61,22 @@ use core::{
 /// [callsite identifier]: super::callsite::Identifier
 pub struct Metadata<'a> {
     /// The name of the span described by this metadata.
-    name: &'static str,
+    name: Cow<'a, str>,
 
     /// The part of the system that the span that this metadata describes
     /// occurred in.
-    target: &'a str,
+    target: Cow<'a, str>,
 
     /// The level of verbosity of the described span.
     level: Level,
 
     /// The name of the Rust module where the span occurred, or `None` if this
     /// could not be determined.
-    module_path: Option<&'a str>,
+    module_path: Option<Cow<'a, str>>,
 
     /// The name of the source code file where the span occurred, or `None` if
     /// this could not be determined.
-    file: Option<&'a str>,
+    file: Option<Cow<'a, str>>,
 
```

Comment:
One fairly major issue with making this change: we've just changed `tracing` and `tracing-core` to support no-std platforms that _don't_ have `liballoc`. Unfortunately, `Cow` is defined in `liballoc`. We'll need to change this to work gracefully on `no_std` platforms without alloc. 

We now have an  "alloc" feature flag for determining whether liballoc is supported,  and APIs that depend on alloc can now require that feature. That should make it possible to conditionally support `Cow` only when alloc is availble, although it does make the API surface a little bit uglier.

Here's what I would recommend:
1. Change the fields to something like this:
   ```rust
   /// ...
   #[cfg(feature = "alloc")]
   name: Cow<'a, str>,
   #[cfg(not(feature = "alloc"))]
   name: &'a str,
   ```
2. We can't change the return type of API functions based on whether or not a feature flag is enabled, since this would make enabling the feature break code in crates that don't enable it. Therefore, we can't have methods like `Metadata::file()` and `Metadata::line()` return `Option<&str>` when alloc is disabled, and `Option<Cow<'a, str>>` when it is enabled. Instead, we'll need two separate APIs: one which is always available but only returns borrowed strings, and another which returns `Cow`s and is only available when the "alloc" feature is enabled.
  
   I would expect the implementation to look something like this:
   ```rust
   impl<'a> Metadata<'a> {
       /// ...
       pub fn file(&'a self) -> Option<&'a str> {
           #[cfg(feature = "alloc")]
           self.file.as_ref().map(Cow::as_ref())
           #[cfg(not(feature = "alloc"))]
           self.file.as_ref()
       }
   
       /// ...
       #[cfg(feature = "alloc")]
       #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
       pub fn file_cow(&self) -> Option<&Cow<'a, String>> {
           self.file.as_ref()
       }
   }
   ```
   (...and so on for the other fields).

   Note that I don't particularly care for the name `file_cow` (and `target_cow` etc), but I don't know if there are better options --- `file_as_cow` feels a little wordy? If you can think of a better naming scheme for these, we can go with that.
3. Finally, I think we'll need two constructors. The `const fn new` will continue to take '&str's only, and always be available, regardless of features. We'll add a new constructor which takes `Cow`s (or `impl Into<Cow<'a, str>>`s?), and have that constructor only be exposed when the "alloc" feature is enabled.

Unfortunately, this makes the API a little less pleasant to work with, but it's the only way to feature flag the use of `Cow` without making a non-additive feature that breaks code when it's enabled...

Reply:
Thank you for the detailed feedback, much appreciated!

I made the change as suggested under 1) and while a bit ugly it works fine.

As for your second point I don't think it's necessarily a problem: as long as we can build a `Metadata` with dynamic data, it's fine to keep returning `&str`. So for example we keep `Metadata::file` returning an `Option<&str>` regardless of how it was built. FWIW I started this work thinking we'd need to duplicate the API and have methods return `Cow`s, but @gnunicorn suggested we just keep things as they were. Do you think there's a need for/value in having the `*_cow()` methods?

I also added a `from_cow` constructor for your third point, featured gated. Struggling a bit to come up with a sensible test for it though.

Reply:
> FWIW I started this work thinking we'd need to duplicate the API and have methods return Cows, but @gnunicorn suggested we just keep things as they were. Do you think there's a need for/value in having the *_cow() methods?

yeah, I really can't think of a good reason, why we'd need to expose the `cow` (in particular if non-mut) over just keeping the current expose functions.

Maybe re 2, I'd have the suggestion of just having two API-equivialent `impl` (and maybe even `struct`) blocks one for when the feature is activated and one without – and the earlier has the extra constructor for non-static-string. I think that would improve legibility quite a bit.

Reply:
>  I think that would improve legibility quite a bit.

I tried this and while the code does look better, I think I would need to duplicate the docs which increases the maintenance burden.

```rust
/// Lots of docs here to maintain
#[cfg(feature = "alloc")]
#[cfg_attr(docsrs, doc(cfg(any(feature = "std", feature = "alloc"))))]
pub struct Metadata<'a> {
}

/// Lots of docs here to maintain
#[cfg(not(feature = "alloc"))]
#[cfg_attr(docsrs, doc(cfg(not(any(feature = "std", feature = "alloc")))))]
pub struct Metadata<'a> {
}

#[cfg(not(feature = "alloc"))]
impl<'a> Metadata<'a> {
    /// Construct new metadata for a span or event, with a name, target, level, field
    /// names, and optional source code location.
    pub const fn new(
        …
    }
}

#[cfg(feature = "alloc")]
impl<'a> Metadata<'a> {
    /// Construct new metadata for a span or event, with a name, target, level, field
    /// names, and optional source code location.
    pub const fn new(…)
        …
    }

    /// Construct new metadata for a span or event, with a name, target, level, field
    /// names, and optional source code location using dynamically allocated data.
    pub fn from_cow(…) {
        …
    }
}

impl<'a> Metadata<'a> {
    // common impls here
}
```

I think it's better to have an easier time maintaining the docs than more readable code in this case, but I'm easy and can do either. 

@hawkw thoughts?